lib/Sema/CSDiag.cpp

```diff
@@ -2063,7 +2063,19 @@ bool FailureDiagnosis::diagnoseContextualConversionError(
   if (contextualType->is<InOutType>())
     options |= TCC_AllowLValue;
 
-  auto *recheckedExpr = typeCheckChildIndependently(expr, options);
+  Expr *recheckedExpr;
+  Type newContextualType = contextualType;
+  do {
+    newContextualType = newContextualType->getSuperclass();
+    if (newContextualType) {
+      recheckedExpr = typeCheckChildIndependently(expr, newContextualType, CTP,
+                                                  options);
+    }
+    else {
+      recheckedExpr = typeCheckChildIndependently(expr, options);
+    }
+  } while(newContextualType && !recheckedExpr && !CS.TC.Diags.hadAnyError());
```

Comment:
Please don't do it this way, it would be much simpler to detect this mismatch in the solver and fix it in `repairFailures`.

Reply:
I'll look into this; just wanted to push my progress.